keps/sig-node/2727-grpc-probe/README.md

```diff
@@ -0,0 +1,168 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in
+  [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture
+      and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+  [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+  links to mailing list discussions/SIG meetings, relevant PRs/issues,
+  release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Goals
+
+Enable grpc probe natively from Kubelet without requiring users to package a
+gRPC healthcheck binary with their container.
+
+https://github.com/grpc-ecosystem/grpc-health-probe
+https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Non-Goals
+
+Add gRPC support in other areas of K8s (e.g. Services).
+
+## Proposal
+
+Host: pod ip
+
+```yaml
+    readinessProbe:
+      grpc:                     #+
+        port: 9090              #+
+        service: my-service     #+
+      initialDelaySeconds: 5
+      periodSeconds: 10
+```
+
+This healthcheck request will be identified with the following gRPC
+`User-Agent` metadata:
+
+```
+User-Agent: kube-probe/K8S_MAJOR_VER.K8S_MINOR_VER
+```
+
+Example:
+
+```
+User-Agent: kube-probe/1.22
+```
+
+### Risks and Mitigations
+
+1. Adds more code to Kubelet and surface area to Pod.Spec. *Response*: we
+   expect that this will be generally useful given broad gRPC adoption in the
+   industry.
+
+## Design Details
+
+```go
+// core/v1/types.go
+
+type Handler struct {
+  // ...
+  TCPSocket *TCPSocketAction `json...`
+
+  // GRPC specifies an action involving a TCP port.          //+
+  // +optional                                               //+
+  GRPC *GRPCAction `json...`                                 //+
+
+  // ...
+}
+
+type GRPCAction struct {
+  // Port name or number of the port to access on the container.
+  // Number must be in the range 1 to 65535.
+  Port intstr.IntOrString `json:"port" protobuf:"bytes,2,opt,name=port"`
```

Comment:
Do we really need names?  That's been such a mess.

Reply:
I'll change to `Int`. It introduces an inconsistency, but that's ok.

Reply:
This came up as a question in KEP reading club. @thockin could you please elaborate on what you mean by port names being a mess?